Broker tries to claim more than 174763 bytes when timing out jobs

I'm assuming this is related to the same issue we had with activating, e.g. something with batching jobs?

Activating a lot of jobs without completing/failing them, with no payload whatsoever, I eventually ran into the following:

```
java.lang.RuntimeException: Unexpected error occurred during RocksDB transaction.
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:169) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:153) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileTrue(ZeebeTransactionDb.java:289) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:76) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.broker.job.JobState.forEachTimedOutEntry(JobState.java:186) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.broker.job.JobTimeoutTrigger.deactivateTimedOutJobs(JobTimeoutTrigger.java:60) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) ~[zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.run(ActorThread.java:108) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
Caused by: java.lang.RuntimeException: Cannot claim more than 174763 bytes.
        at io.zeebe.dispatcher.Dispatcher.claim(Dispatcher.java:294) ~[zb-dispatcher-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.logstreams.log.LogStreamBatchWriterImpl.claimBatchForEvents(LogStreamBatchWriterImpl.java:233) ~[zb-logstreams-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.logstreams.log.LogStreamBatchWriterImpl.tryWrite(LogStreamBatchWriterImpl.java:211) ~[zb-logstreams-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.broker.logstreams.processor.TypedCommandWriterImpl.flush(TypedCommandWriterImpl.java:143) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.broker.job.JobTimeoutTrigger.lambda$deactivateTimedOutJobs$1(JobTimeoutTrigger.java:65) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.broker.job.JobState.lambda$forEachTimedOutEntry$6(JobState.java:198) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.visit(ZeebeTransactionDb.java:385) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileTrue$7(ZeebeTransactionDb.java:295) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.runInNewTransaction(ZeebeTransactionDb.java:177) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:166) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
```

Interesting lines are the following:

>at **io.zeebe.broker.job.JobTimeoutTrigger.lambda$deactivateTimedOutJobs$1(JobTimeoutTrigger.java:65)** ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
> at **io.zeebe.broker.job.JobState.lambda$forEachTimedOutEntry$6(JobState.java:198)** ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]

Happened to Bernd also, we should investigate the root cause

```
22:54:42.358 [zb-stream-processor] [10.244.0.170:26501-zb-actors-1] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.RuntimeException: Unexpected error occurred during RocksDB transaction.
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:169) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:153) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileTrue(ZeebeTransactionDb.java:289) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:76) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.broker.job.JobState.forEachTimedOutEntry(JobState.java:186) ~[zeebe-broker-core-0.16.1.jar:0.16.1]
	at io.zeebe.broker.job.JobTimeoutTrigger.deactivateTimedOutJobs(JobTimeoutTrigger.java:60) ~[zeebe-broker-core-0.16.1.jar:0.16.1]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) ~[zb-util-0.16.1.jar:0.16.1]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zb-util-0.16.1.jar:0.16.1]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [zb-util-0.16.1.jar:0.16.1]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zb-util-0.16.1.jar:0.16.1]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zb-util-0.16.1.jar:0.16.1]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [zb-util-0.16.1.jar:0.16.1]
Caused by: java.lang.RuntimeException: Cannot claim more than 174763 bytes.
	at io.zeebe.dispatcher.Dispatcher.claim(Dispatcher.java:294) ~[zb-dispatcher-0.16.1.jar:0.16.1]
	at io.zeebe.logstreams.log.LogStreamBatchWriterImpl.claimBatchForEvents(LogStreamBatchWriterImpl.java:233) ~[zb-logstreams-0.16.1.jar:0.16.1]
	at io.zeebe.logstreams.log.LogStreamBatchWriterImpl.tryWrite(LogStreamBatchWriterImpl.java:211) ~[zb-logstreams-0.16.1.jar:0.16.1]
	at io.zeebe.broker.logstreams.processor.TypedCommandWriterImpl.flush(TypedCommandWriterImpl.java:143) ~[zeebe-broker-core-0.16.1.jar:0.16.1]
	at io.zeebe.broker.job.JobTimeoutTrigger.lambda$deactivateTimedOutJobs$1(JobTimeoutTrigger.java:65) ~[zeebe-broker-core-0.16.1.jar:0.16.1]
	at io.zeebe.broker.job.JobState.lambda$forEachTimedOutEntry$6(JobState.java:198) ~[zeebe-broker-core-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.visit(ZeebeTransactionDb.java:385) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileTrue$7(ZeebeTransactionDb.java:295) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.runInNewTransaction(ZeebeTransactionDb.java:177) ~[zb-db-0.16.1.jar:0.16.1]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:166) ~[zb-db-0.16.1.jar:0.16.1]
	... 11 more
```

--- a/broker-core/src/main/java/io/zeebe/broker/job/JobState.java
+++ b/broker-core/src/main/java/io/zeebe/broker/job/JobState.java
@@ -28,7 +28,6 @@ import io.zeebe.db.impl.DbNil;
 import io.zeebe.db.impl.DbString;
 import io.zeebe.protocol.impl.record.value.job.JobRecord;
 import io.zeebe.util.EnsureUtil;
-import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import org.agrona.DirectBuffer;
 
@@ -181,7 +180,7 @@ public class JobState {
   }
 
   public void forEachTimedOutEntry(
-      final long upperBound, final BiConsumer<Long, JobRecord> callback) {
+      final long upperBound, final BiFunction<Long, JobRecord, Boolean> callback) {
 
     deadlinesColumnFamily.whileTrue(
         (compositeKey, zbNil) -> {
@@ -189,15 +188,9 @@ public class JobState {
           final boolean isDue = deadline < upperBound;
           if (isDue) {
             final long jobKey = compositeKey.getSecond().getValue();
-            final JobRecord job = getJob(jobKey);
-
-            if (job == null) {
-              throw new IllegalStateException(
-                  String.format("Expected to find job with key %d, but no job found", jobKey));
-            }
-            callback.accept(jobKey, job);
+            return visitJob(jobKey, callback);
           }
-          return isDue;
+          return false;
         });
   }
 
@@ -230,14 +223,17 @@ public class JobState {
         jobTypeKey,
         ((compositeKey, zbNil) -> {
           final long jobKey = compositeKey.getSecond().getValue();
-
-          final JobRecord job = getJob(jobKey);
-          if (job == null) {
-            throw new IllegalStateException(
-                String.format("Expected to find job with key %d, but no job found", jobKey));
-          }
-          return callback.apply(jobKey, job);
+          return visitJob(jobKey, callback);
         }));
+  }
+
+  private boolean visitJob(long jobKey, BiFunction<Long, JobRecord, Boolean> callback) {
+    final JobRecord job = getJob(jobKey);
+    if (job == null) {
+      throw new IllegalStateException(
+          String.format("Expected to find job with key %d, but no job found", jobKey));
+    }
+    return callback.apply(jobKey, job);
   }
 
   public JobRecord updateJobRetries(final long jobKey, final int retries) {

--- a/broker-core/src/main/java/io/zeebe/broker/job/JobTimeoutTrigger.java
+++ b/broker-core/src/main/java/io/zeebe/broker/job/JobTimeoutTrigger.java
@@ -55,16 +55,19 @@ public class JobTimeoutTrigger implements StreamProcessorLifecycleAware {
     }
   }
 
-  private void deactivateTimedOutJobs() {
+  void deactivateTimedOutJobs() {
     final long now = currentTimeMillis();
     state.forEachTimedOutEntry(
         now,
         (key, record) -> {
           writer.appendFollowUpCommand(
               key, JobIntent.TIME_OUT, record, (m) -> m.valueType(ValueType.JOB));
-          writer.flush();
-          // we don't have to check for write errors as the job will then be picked up again
-          // on the next iteration of this timer
+
+          final boolean flushed = writer.flush() >= 0;
+          if (!flushed) {
+            writer.reset();
+          }
+          return flushed;
         });
   }
 }

--- a/broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedCommandWriter.java
+++ b/broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedCommandWriter.java
@@ -32,6 +32,8 @@ public interface TypedCommandWriter {
   void appendFollowUpCommand(
       long key, Intent intent, UnpackedObject value, Consumer<RecordMetadata> metadata);
 
+  void reset();
+
   /** @return position of new record, negative value on failure */
   long flush();
 }

--- a/broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedCommandWriterImpl.java
+++ b/broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedCommandWriterImpl.java
@@ -134,6 +134,7 @@ public class TypedCommandWriterImpl implements TypedCommandWriter {
     appendRecord(key, RecordType.COMMAND, intent, value, metadata);
   }
 
+  @Override
   public void reset() {
     batchWriter.reset();
   }

--- a/broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedStreamWriter.java
+++ b/broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedStreamWriter.java
@@ -43,7 +43,4 @@ public interface TypedStreamWriter extends TypedCommandWriter {
       long key, Intent intent, UnpackedObject value, Consumer<RecordMetadata> metadata);
 
   KeyGenerator getKeyGenerator();
-
-  /** @return position of new event, negative value on failure */
-  long flush();
 }

--- a/broker-core/src/test/java/io/zeebe/broker/job/JobStateTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/job/JobStateTest.java
@@ -235,6 +235,30 @@ public class JobStateTest {
   }
 
   @Test
+  public void shouldOnlyIterateOverTimedoutWhileTrue() {
+    // given
+    createAndActivateJobRecord(1, newJobRecord().setDeadline(1L));
+    createAndActivateJobRecord(2, newJobRecord().setDeadline(256L));
+    createAndActivateJobRecord(3, newJobRecord().setDeadline(512L));
+    createAndActivateJobRecord(4, newJobRecord().setDeadline(65536L));
+    createAndActivateJobRecord(5, newJobRecord().setDeadline(4294967296L));
+
+    // when
+    final List<Long> timedOutKeys = new ArrayList<>();
+    final long since = 65536L;
+    jobState.forEachTimedOutEntry(
+        since,
+        (k, e) -> {
+          timedOutKeys.add(k);
+          return k.longValue() < 3;
+        });
+
+    // then
+    assertThat(timedOutKeys).hasSize(3);
+    assertThat(timedOutKeys).containsExactly(1L, 2L, 3L);
+  }
+
+  @Test
   public void shouldDoNothingIfNotTimedOutJobs() {
     // given
     jobState.create(5, newJobRecord().setDeadline(512L));

--- a/broker-core/src/test/java/io/zeebe/broker/job/JobTimeoutTriggerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/job/JobTimeoutTriggerTest.java
@@ -1,0 +1,101 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.job;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import io.zeebe.broker.logstreams.processor.TypedStreamEnvironment;
+import io.zeebe.broker.logstreams.processor.TypedStreamProcessor;
+import io.zeebe.broker.logstreams.processor.TypedStreamWriter;
+import io.zeebe.broker.util.ZeebeStateRule;
+import io.zeebe.protocol.impl.record.value.job.JobRecord;
+import io.zeebe.protocol.intent.JobIntent;
+import io.zeebe.util.sched.ActorControl;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+
+public class JobTimeoutTriggerTest {
+
+  @Rule public final ZeebeStateRule stateRule = new ZeebeStateRule();
+
+  @Mock private ActorControl someActor;
+
+  @Mock private TypedStreamWriter typedStreamWriter;
+  private JobTimeoutTrigger jobTimeoutTrigger;
+
+  @Before
+  public void setUp() {
+    initMocks(this);
+
+    final JobState jobState = stateRule.getZeebeState().getJobState();
+    jobTimeoutTrigger = new JobTimeoutTrigger(jobState);
+
+    final TypedStreamProcessor streamProcessor = mock(TypedStreamProcessor.class);
+    when(streamProcessor.getActor()).thenReturn(someActor);
+    final TypedStreamEnvironment environment = mock(TypedStreamEnvironment.class);
+    when(environment.buildCommandWriter()).thenReturn(typedStreamWriter);
+    when(streamProcessor.getEnvironment()).thenReturn(environment);
+
+    jobTimeoutTrigger.onRecovered(streamProcessor);
+
+    jobState.activate(0, newJobRecord());
+    jobState.activate(1, newJobRecord());
+    jobState.activate(2, newJobRecord());
+  }
+
+  private JobRecord newJobRecord() {
+    final JobRecord jobRecord = new JobRecord();
+
+    jobRecord.setRetries(2);
+    jobRecord.setDeadline(256L);
+    jobRecord.setType("test");
+
+    return jobRecord;
+  }
+
+  @Test
+  public void shouldNotWriteAgainAfterFlushFailed() {
+    // given
+    when(typedStreamWriter.flush()).thenReturn(1L, -1L);
+
+    // when
+    jobTimeoutTrigger.deactivateTimedOutJobs();
+
+    // then
+    final InOrder inOrder = Mockito.inOrder(typedStreamWriter);
+
+    inOrder
+        .verify(typedStreamWriter)
+        .appendFollowUpCommand(eq(0L), eq(JobIntent.TIME_OUT), any(JobRecord.class), any());
+    inOrder.verify(typedStreamWriter).flush();
+    inOrder
+        .verify(typedStreamWriter)
+        .appendFollowUpCommand(eq(1L), eq(JobIntent.TIME_OUT), any(JobRecord.class), any());
+    inOrder.verify(typedStreamWriter).flush();
+    inOrder.verify(typedStreamWriter).reset();
+    inOrder.verifyNoMoreInteractions();
+  }
+}

